pkg/analysis_server/lib/src/computer/computer_closingLabels.dart

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analysis_server/protocol/protocol_generated.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/src/generated/source.dart';
+
+/**
+ * A computer for [CompilationUnit] closing labels.
+ */
+class DartUnitClosingLabelsComputer {
+  final LineInfo _lineInfo;
+  final CompilationUnit _unit;
+  final List<ClosingLabel> _closingLabels = <ClosingLabel>[];
+
+  DartUnitClosingLabelsComputer(this._lineInfo, this._unit);
+
+  /**
+   * Returns a list of closing labels, not `null`.
+   */
+  List<ClosingLabel> compute() {
+    _unit.accept(new _DartUnitClosingLabelsComputerVisitor(this));
+    return _closingLabels;
+  }
+}
+
+/**
+ * An AST visitor for [DartUnitClosingLabelsComputer].
+ */
+class _DartUnitClosingLabelsComputerVisitor
+    extends RecursiveAstVisitor<Object> {
+  final DartUnitClosingLabelsComputer computer;
+
+  _DartUnitClosingLabelsComputerVisitor(this.computer);
+
+  @override
+  Object visitInstanceCreationExpression(InstanceCreationExpression node) {
+    if (_spansManyLines(node)) {
+      // TODO(dantup) Is it right to read constructorName to get nameless constructors types?
```

Comment:
It won't necessarily give you the actual name of the class of object that will be returned (in the case of a factory constructor), but it will match what's in the code, so I'd say yes.